pkg/apis/crd/v1beta1/types.go

```diff
@@ -243,3 +243,259 @@ type IPPoolUsage struct {
 	// Number of allocated IPs.
 	Used int `json:"used"`
 }
+
+type TraceflowPhase string
+
+const (
+	Running   TraceflowPhase = "Running"
+	Succeeded TraceflowPhase = "Succeeded"
+	Failed    TraceflowPhase = "Failed"
+)
+
+type TraceflowComponent string
+
+const (
+	ComponentSpoofGuard    TraceflowComponent = "SpoofGuard"
+	ComponentLB            TraceflowComponent = "LB"
+	ComponentRouting       TraceflowComponent = "Routing"
+	ComponentNetworkPolicy TraceflowComponent = "NetworkPolicy"
+	ComponentForwarding    TraceflowComponent = "Forwarding"
+	ComponentEgress        TraceflowComponent = "Egress"
+)
+
+type TraceflowAction string
+
+const (
+	ActionDelivered TraceflowAction = "Delivered"
+	ActionReceived  TraceflowAction = "Received"
+	ActionForwarded TraceflowAction = "Forwarded"
+	ActionDropped   TraceflowAction = "Dropped"
+	ActionRejected  TraceflowAction = "Rejected"
+	// ActionForwardedOutOfOverlay indicates that the packet has been forwarded out of the network
+	// managed by Antrea. This indicates that the Traceflow request can be considered complete.
+	ActionForwardedOutOfOverlay TraceflowAction = "ForwardedOutOfOverlay"
+	ActionMarkedForSNAT         TraceflowAction = "MarkedForSNAT"
+	ActionForwardedToEgressNode TraceflowAction = "ForwardedToEgressNode"
+)
+
+// List the supported protocols and their codes in traceflow.
+// According to code in Antrea agent and controller, default protocol is ICMP if protocol is not inputted by users.
+const (
+	ICMPProtocolNumber int32 = 1
+	IGMPProtocolNumber int32 = 2
+	TCPProtocolNumber  int32 = 6
+	UDPProtocolNumber  int32 = 17
+	SCTPProtocolNumber int32 = 132
+)
+
+var SupportedProtocols = map[string]int32{
+	"TCP":  TCPProtocolNumber,
+	"UDP":  UDPProtocolNumber,
+	"ICMP": ICMPProtocolNumber,
+}
+
+var ProtocolsToString = map[int32]string{
+	TCPProtocolNumber:  "TCP",
+	UDPProtocolNumber:  "UDP",
+	ICMPProtocolNumber: "ICMP",
+	IGMPProtocolNumber: "IGMP",
+	SCTPProtocolNumber: "SCTP",
+}
+
+// List the supported destination types in traceflow.
+const (
+	DstTypePod     = "Pod"
+	DstTypeService = "Service"
+	DstTypeIPv4    = "IPv4"
+)
+
+var SupportedDestinationTypes = []string{
+	DstTypePod,
+	DstTypeService,
+	DstTypeIPv4,
+}
+
+// Default timeout in seconds.
+const DefaultTraceflowTimeout int32 = 20
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   TraceflowSpec   `json:"spec,omitempty"`
+	Status TraceflowStatus `json:"status,omitempty"`
+}
+
+// TraceflowSpec describes the spec of the traceflow.
+type TraceflowSpec struct {
+	Source      Source      `json:"source,omitempty"`
```

Comment:
Should all Source, Destination, Packet be pointers?

Reply:
> Should all Source, Destination, Packet be pointers?

The pros and cons of setting the types to pointers:

- Pros: we can distinguish whether the field is specified by users or not because we allow nil value.
- Cons: we have to check whether this field is nil or not every time we access it. If we forget, a panic will be raised.

Reply:
I suppose all of them including TransportHeader can be pointers, there are indeed some pros and cons like @shi0rik0 mentioned above. Let me know if you feel it's better to be pointers. I will also double check with @gran-vmv in case I missed anything for the original design.

Reply:
If we needs to check individual fields in a struct to know the struct is specified by users, I think a pointer would be better.

Reply:
I would like to defer related changes to next release considering we may need more efforts for verification.
@shi0rik0 please check if you can take care of this when you make the firstN sampling changes, including those suggestions you raise in the comments. Thanks.

Reply:
I don't have a strong opinion about changing these fields to pointers, but if we do make that change, it should be done in its own PR (not when implementing a new feature like firstN sampling).